### PR TITLE
Convert project to TypeScript

### PR DESCRIPTION
**Pull request by benhjames**
Follows the conventions in the other Helmet repos.

This should operate in exactly the same way as before, with the exception of the arguments length check. I removed this check as I thought it wasn't adding much, and was making TypeScript complain about the arguments length test in the test script. Note that this wouldn't be a breaking change.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import { IncomingMessage, ServerResponse } from 'http';
+import depd from 'depd';
+
+const deprecate = depd('hsts');
+
+const DEFAULT_MAX_AGE = 180 * 24 * 60 * 60;
+
+interface HSTSOptions {
```
Comment (EvanHahn) [inline index.ts]: Nit: this should be camel-cased as `HstsOptions`.

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import { IncomingMessage, ServerResponse } from 'http';
+import depd from 'depd';
+
+const deprecate = depd('hsts');
```
Comment (EvanHahn) [inline index.ts]: Why is this being double-imported?

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  }
+  if (options.preload) {
+    header += '; preload';
+  }
+
+  return function hsts (req: IncomingMessage, res: ServerResponse, next: () => void) {
+    if (setIf(req, res)) {
+      res.setHeader('Strict-Transport-Security', header);
+    }
+
```
Comment (EvanHahn) [inline index.ts]: Nit: could you remove this line break?

File: index.ts (hunk 2 above)
Comment (benhjames) [reply index.ts]: This matches the previous behaviour:

```js
var deprecate = require('depd')('hsts')
```

Comment (EvanHahn) [reply index.ts]: Ah, apologies—I misread. Resolving this thread.

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+export = function hsts (options: HstsOptions = {}) {
+  if ('includeSubdomains' in options) {
+    deprecate('The "includeSubdomains" parameter is deprecated. Use "includeSubDomains" (with a capital D) instead.');
+  }
+
+  if ('setIf' in options) {
+    deprecate('The "setIf" parameter is deprecated. Refer to the documentation to see how to set the header conditionally.');
+  }
+
+  const { maxAge = DEFAULT_MAX_AGE, setIf = alwaysTrue } = options;
```
Comment (EvanHahn) [inline index.ts]: The original behavior isn't ideal, but this is actually a breaking change. The old line looked like this:

```js
var maxAge = options.maxAge != null ? options.maxAge : DEFAULT_MAX_AGE
```

That means that the a parameter of `{}` and `{maxAge: null}` were equivalent. Now, they're not. I filed #26 to fix it when we next make a breaking change, but we should fix the behavior here.

File: index.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  if ('maxage' in options) {
```
Comment (EvanHahn) [inline index.ts]: Can we revert this to `options.hasOwnProperty('maxage')`?

File: index.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  if (typeof maxAge !== 'number') {
+    throw new TypeError('HSTS must be passed a numeric maxAge parameter.');
+  }
+  if (maxAge < 0) {
+    throw new RangeError('HSTS maxAge must be nonnegative.');
+  }
+  if (typeof setIf !== 'function') {
+    throw new TypeError('setIf must be a function.');
+  }
+  if ('includeSubDomains' in options && 'includeSubdomains' in options) {
```
Comment (EvanHahn) [inline index.ts]: Similar comment to the above—could we revert this to `options.hasOwnProperty` instead?

File: package-lock.json [hunk 7]
```diff
@@ -4,1133 +4,3615 @@
...
-    "acorn-jsx": {
-      "version": "5.0.1",
-      "resolved": "https://registry.npmjs.org/acorn-jsx/-/acorn-jsx-5.0.1.tgz",
-      "integrity": "sha512-HJ7CfNHrfJLlNTzIEUTj43LNWGkqpRLxm3YjAlcD0ACydk9XynzYsCBHxut+iqt+1aBXkx9UP/w/ZqMr13XIzg==",
-      "dev": true
+    "@babel/code-frame": {
+      "version": "7.0.0",
+      "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.0.0.tgz",
+      "integrity": "sha512-OfC2uemaknXr87bdLUkWog7nYuliM9Ij5HUcajsVcMCpQrcLmtxRbVFTIqmcSkSeYRBFBRxs2FiUqFJDLdiebA==",
+      "dev": true,
```
[the remark this answers is not in the document]
Comment (benhjames) [reply package-lock.json]: Have upgraded npm to `6.9.0` and updated the lockfile 👍

File: index.ts (hunk 5 above)
Comment (benhjames) [reply index.ts]: Sure, though note the verbosity of `Object.prototype.hasOwnProperty.call(..., ...)` due to [this eslint rule](https://eslint.org/docs/rules/no-prototype-builtins)

File: test/index.test.ts [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    const server = app(hsts({
+      setIf(req) {
+        return req.headers['x-should-set'] === 'yes';
+      },
+    }));
+
+    const shouldntSetPromise = request(server)
+      .get('/')
+      .set('X-Should-Set', 'no')
+      .then((res: any) => {
```
Comment (EvanHahn) [inline test/index.test.ts]: I get a lint warning here. Could you fix this?

File: package.json [hunk 9]
```diff
@@ -27,24 +27,38 @@
...
+    "prepublishOnly": "npm run build",
+    "lint": "eslint --fix '**/*.ts'",
+    "test": "jest --config test/jest-config.json",
+    "clean": "rm -rf dist",
+    "build": "npm run clean && tsc"
   },
   "dependencies": {
     "depd": "2.0.0"
+  },
+  "devDependencies": {
```
Comment (EvanHahn) [inline package.json]: `npm outdated` showed that a few of these deps were out of date. Could you update @types/jest, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, and typescript to their latest versions?

`npm install @types/jest@latest @typescript-eslint/eslint-plugin/latest @typescript-eslint/parser@latest typescript@latest --save-dev` should do it.

**Comment by EvanHahn** [comment]
Thanks! I'll deploy this soon.

**Pull request merged**